provider/cloudflare.go

```diff
@@ -0,0 +1,278 @@
+				recordID, err := p.getRecordID(zoneID, change.ResourceRecordSet)
+				if err != nil {
+					log.WithFields(logFields).Errorf("Failed to get record: %v", err)
+				}
+				err = p.Client.DeleteDNSRecord(zoneID, recordID)
+				if err != nil {
+					log.WithFields(logFields).Errorf("Failed to delete record: %v", err)
+				}
+			case cloudFlareUpdate:
+				recordID, err := p.getRecordID(zoneID, change.ResourceRecordSet)
+				if err != nil {
+					log.WithFields(logFields).Errorf("Failed to get record: %v", err)
+				}
+				err = p.Client.UpdateDNSRecord(zoneID, recordID, change.ResourceRecordSet)
+				if err != nil {
+					log.WithFields(logFields).Errorf("Failed to update record: %v", err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// changesByZone separates a multi-zone change into a single change per zone.
+func cloudflareChangesByZone(zones []cloudflare.Zone, changeSet []*cloudFlareChange) map[string][]*cloudFlareChange {
+	changes := make(map[string][]*cloudFlareChange)
+
+	for _, z := range zones {
+		changes[z.ID] = []*cloudFlareChange{}
+	}
+
+	for _, c := range changeSet {
+		zone := cloudflareSuitableZone(c.ResourceRecordSet.Name, zones)
+		if zone == nil {
+			log.Debugf("Skipping record %s because no hosted zone matching record DNS Name was detected ", c.ResourceRecordSet.Name)
+			continue
+		}
+		changes[zone.ID] = append(changes[zone.ID], c)
+	}
+
+	return changes
+}
+
+// cloudflareSuitableZone returns the most suitable zone for a given hostname
+// and a set of zones.
+func cloudflareSuitableZone(hostname string, zones []cloudflare.Zone) *cloudflare.Zone {
+	var zone cloudflare.Zone
+	for _, z := range zones {
+		if strings.HasSuffix(hostname, z.Name) {
+			if len(z.Name) > len(zone.Name) {
+				zone = z
+			}
+		}
+	}
+	return &zone
+}
+
+func (p *cloudFlareProvider) getRecordID(zoneID string, record cloudflare.DNSRecord) (string, error) {
+	records := cloudflare.DNSRecord{}
+	zoneRecords, err := p.Client.DNSRecords(zoneID, records)
+	if err != nil {
+		return "", err
+	}
+	for _, zoneRecord := range zoneRecords {
+		if zoneRecord.Name == record.Name {
+			return zoneRecord.ID, nil
+		}
+	}
+	return "", fmt.Errorf("No record id found")
+}
+
+// newCloudFlareChanges returns a collection of Changes based on the given records and action.
+func newCloudFlareChanges(action string, endpoints []*endpoint.Endpoint) []*cloudFlareChange {
+	changes := make([]*cloudFlareChange, 0, len(endpoints))
+
+	for _, endpoint := range endpoints {
+		changes = append(changes, newCloudFlareChange(action, endpoint))
+	}
+
+	return changes
+}
+
+func newCloudFlareChange(action string, endpoint *endpoint.Endpoint) *cloudFlareChange {
+	typ := suitableType(endpoint)
+
+	return &cloudFlareChange{
+		Action: action,
+		ResourceRecordSet: cloudflare.DNSRecord{
+			Name: endpoint.DNSName,
+			// TTL Value of 1 is 'automatic'
+			TTL: 1,
+			// record is receiving the performance and security benefits of CloudFlare
+			Proxied: typ == "A" || typ == "CNAME",
```

Comment:
We talked about making this `false` unless there is like a global config were you can define this.

Ideally it should be possible to configure per resource, but I think it's out of scope for this PR.